barretenberg/cpp/pil/vm2/tx.pil

```diff
@@ -0,0 +1,343 @@
+    pol commit is_revertible;
+    pol commit reverted;
+    // phase_value stays the same unless we are at end_phase or reverted
+    #[PHASE_VALUE_CONTINUITY]
+    NOT_PHASE_END * (1 - reverted) * (1 - precomputed.first_row) * (phase_value' - phase_value) = 0;
+    // At the start of a new phase, we need to increment the phase value
+    #[INCR_PHASE_VALUE_ON_END]
+    NOT_LAST * (1 - reverted) * end_phase * (phase_value' - (phase_value + 1)) = 0;
+
+    // If reverted == 1, is_revertible must be 1
+    reverted * (1 - is_revertible) = 0;
+
+    // Control flow
+    pol commit read_pi_offset; // index in public inputs to read data from
+    pol commit write_pi_offset; // index in public inputs to write data to
+    pol commit read_pi_length_offset; // index in public inputs to read lengths from
+
+    pol commit remaining_phase_inv;
+    pol commit remaining_phase_counter;
+    pol commit remaining_phase_minus_one_inv;
+    // If remaining_phase_counter == 0, is_padded = 1
+    #[REM_COUNT_IS_ZERO]
+    sel * (remaining_phase_counter * (is_padded * (1 - remaining_phase_inv) + remaining_phase_inv) - 1 + is_padded) = 0;
+
+    pol REM_COUNT_MINUS_1 = remaining_phase_counter - 1;
+    // If remaining_phase_counter == 1, end_phase = 1
+    #[REM_COUNT_IS_ONE]
+    sel * (1 - is_padded) * (REM_COUNT_MINUS_1 * (end_phase * (1 - remaining_phase_minus_one_inv) + remaining_phase_minus_one_inv) - 1 + end_phase) = 0;
+    
+    #[READ_PHASE_TABLE]
+    start_phase { 
+        phase_value, 
+        is_public_call_request, 
+        is_l2_l1_msg_phase,
+        /* Tree Padding Phase */
+        is_collect_fee,
+
+        is_revertible,
+        read_pi_offset,
+        read_pi_length_offset,
+        write_pi_offset,
+        sel_non_revertible_append_note_hash,
+        sel_non_revertible_append_nullifier,
+        sel_revertible_append_note_hash,
+        sel_revertible_append_nullifier
+    }
+    in
+    precomputed.phase_sel {
+        precomputed.phase_value, 
+        precomputed.is_public_call_request_phase,
+        precomputed.is_l2_l1_message_phase,
+        /* Tree Padding Phase */
+        precomputed.sel_collect_fee,
+
+        precomputed.is_revertible,
+        precomputed.read_public_input_offset,
+        precomputed.read_public_input_length_offset,
+        precomputed.write_public_input_offset,
+        precomputed.sel_non_revertible_append_note_hash,
+        precomputed.sel_non_revertible_append_nullifier,
+        precomputed.sel_revertible_append_note_hash,
+        precomputed.sel_revertible_append_nullifier
+    };
+
+    // If a phase reverts (when it is allowed to), it jumps to a known next phase
+    #[PHASE_JUMP_ON_REVERT]
+    reverted { phase_value, phase_value' } in
+    precomputed.phase_sel { precomputed.phase_value, precomputed.next_phase_on_revert};
+
+    // === Phase Lengths === 
+    // Read the number of steps in the phase from public inputs
+    pol commit read_phase_length_sel;
```

Comment:
selectors _start_ with `sel_`

(/me inserts bernie saunders meme "I'm once again telling you..." XD)